### ISSUE TITLE
 Error in data.frame(sp, mod_name, rmse, rsq): arguments imply differing number of rows: 0, 1

### ISSUE THREAD

**Issue by dosshra** [nickfountainjones/mrIML#4]
Hello
I tried  to run mrIML using this script:
`cl <- parallel::makeCluster(20)`
   `future::plan(cluster, workers=cl)`
   `X<-as.data.frame(gfdf[,2:13])`
   `Y<-gfdf[,seq(14,61822,100)]`
`model_rf <- 
  rand_forest(trees = 10, mode = "regression", mtry = tune(), min_n = tune()) %>% 
  set_engine("randomForest")`
`yhats_rf <- mrIMLpredicts(X=X,Y=Y, Model=model_rf, balance_data='no', mode='regression', tune_grid_size=5, seed = sample.int(1e8, 1) )`
I get a lot of warning during the run:

> ! Fold05: preprocessor 1/1, model 1/5: The response has five or fewer unique value
>! Fold04: internal: A correlation computation is required, but `truth` is constant

When I run the following code:

`ModelPerf <- mrIMLperformance(yhats_rf, Model=model_rf, Y, mode='regression')`
I get this error:

> Error in data.frame(sp, mod_name, rmse, rsq): arguments imply differing number of rows: 0, 1
Traceback:

>1. mrIMLperformance(yhats_rf, Model = model_rf, Y, mode = "regression")
>2. data.frame(sp, mod_name, rmse, rsq)
>3. stop(gettextf("arguments imply differing number of rows: %s", 
> .     paste(unique(nrows), collapse = ", ")), domain = NA)

Thank you

**Comment by nfj1380** [comment]
Hi there,

Can you provide a reproducible example? I suspect it is to do with the Y variables here.
 The warnings you get are common during the tuning process - it means for some tuning parameters in the models there wasn't enough signal to calculate r2 properly. Usually these warnings aren't a huge issue.

Cheers,

Nick

**Comment by dosshra** [comment]
Thank you for the reply. 
Please see a RData file with X&Y. Y is a matrix of   minor allele proportions in 110 populations, and X is the environmental variables of the populations. 
[mrIML.zip](https://github.com/nfj1380/mrIML/files/8281359/mrIML.zip)

**Comment by nfj1380** [comment]
I couldn't recreate your error - mrIMLperformance worked fine for me with your data. When did you install the package?

**Comment by dosshra** [comment]
I found a problem with my Y matrix.
It is running OK now. Thank you

**Comment by dosshra** [comment]
Hi
Using the data that I uploaded,  and the code in the original message, It takes 4 hours with 20 CPU to complete. Y is only 110x618 matrix, X includes 12 variables, and I run only 10 trees.  
The  mrIML installation is from last week on R version 3.6.3. Trying with R 4.1.2 with 7 CPU also took a very long time. What could be the problem?  
Thank you

**Comment by nfj1380** [comment]
You could try to reduce the tuning grid size (i.e. not test as many hyperparameter combinations). If you are comparing to gradient forests this is why it takes longer - GF doesn't do any model tuning. That said,  it took overnight on my machine using 6 cores which is longer than I thought it would given the data dimensions.You could try other algorithms too  - SVM could be faster if you are on a time crunch. https://parsnip.tidymodels.org/articles/articles/Models.html.
[https://parsnip.tidymodels.org/logo.png]<https://parsnip.tidymodels.org/articles/articles/Models.html>
List of Models • parsnip<https://parsnip.tidymodels.org/articles/articles/Models.html>
parsnip is a part of the tidymodels ecosystem, a collection of modeling packages designed with common APIs and a shared philosophy.
parsnip.tidymodels.org

________________________________

**Comment by nfj1380** [comment]
I just tried with xgb and it worked much faster than the rf model for some reason. Might be also be worth a try.

try this:

model_xgb<-  boost_tree(
  trees = 100,
  tree_depth = tune(), min_n = tune(),
  loss_reduction = tune(),                     ## first three: model complexity
  sample_size = tune(), mtry = tune(),         ## randomness
  learn_rate = tune(),                         ## step size
) %>%
  set_engine("xgboost") %>%
  set_mode("regression")
________________________________

**Comment by dosshra** [comment]
Thank you for the suggestion, xgb do run faster. However: trying to continue exploring the data using this code:
`yhats_rf <- mrIMLpredicts(X=X,Y=Y, Model=model_rf, balance_data='no', mode='regression', tune_grid_size=5, seed = sample.int(1e8, 1) )
ModelPerf <- mrIMLperformance(yhats_rf, Model=model_rf, Y, mode='regression')
VI <- mrVip(yhats=yhats_rf, X=X) ` 
resulted in error.

**Comment by dosshra** [comment]
HI
I was able to run the RF regression for the whole dateset of >60k SNP using `mtry=4, trees = 10 .` It took 6 hrs with 20 CPU and more than 130GB of RAM.
However, the flashlight analysis:
`flashlightObj <- mrFlashlight(yhats_rf, X, Y, response = "multi", mode='regression')
profileData_pd <- light_profile(flashlightObj, v = "november") #partial dependencies
mrProfileplot(profileData_pd , sdthresh =0.1)`
Did not finish after 24 hours.
Thank you

**Comment by nfj1380** [comment]
I think that would be close to a record for number of SNPs in a mrIML analysis  -I'm glad it worked even if it was intensive.

With the VI problem can you provide more information? Also, I have updated that code recently - so try reinstalling from github.

Cheers,

Nick
________________________________
